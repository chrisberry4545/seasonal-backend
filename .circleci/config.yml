version: 2
jobs:
  test:
    docker:
      - image: circleci/node:8.10

    steps:
      - checkout

      - setup_remote_docker

      - run:
          name: Output dir
          command: echo $PWD
      - run:
          name: Run Docker
          command: docker-compose run app npm test

      - run:
          name: Run tests
          command: docker exec app npm run test

  deploy:
    docker:
      - image: circleci/node:8.10

    steps:
      - checkout

      - restore_cache:
          keys:
          - v1-dependencies-
          # fallback to using the latest cache if no exact match is found
          - v1-dependencies-

      - run:
          name: Copy env config
          command: |
            cp .env-ci.yml .env.yml

      - run:
          name: Install Serverless CLI and dependencies
          command: |
            sudo npm i -g serverless
            npm install

      - run:
          name: Run tests
          command: npm test

      - run:
          name: Build dist version
          command: |
            npm run build
            npm run move-dist-to-top-level

      - save_cache:
          paths:
            - node_modules
          key: v1-dependencies-

      - run:
          name: Deploy application
          command: sls deploy --stage prod

workflows:
  version: 2
  ci:
    jobs:
      - deploy:
          filters:
            branches:
              only:
                - master
      - test
