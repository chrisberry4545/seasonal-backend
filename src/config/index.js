module.exports = {
  AIRTABLE_API_KEY: process.env.AIRTABLE_API_KEY,
  SKIP_CACHE: process.env.SKIP_CACHE || false
};
