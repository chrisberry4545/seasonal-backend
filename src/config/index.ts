export const AIRTABLE_API_KEY = process.env.AIRTABLE_API_KEY;
export const AIRTABLE_BASE_DB =
  process.env.AIRTABLE_BASE_DB || 'appCYPaLSsUK9Tp3a';
export const SKIP_CACHE = process.env.SKIP_CACHE || false;
export const SEASON_DATA_ENDPOINT =
  process.env.SEASON_DATA_ENDPOINT || 'season-data';
export const FOOD_DATA_ENDPOINT =
  process.env.FOOD_DATA_ENDPOINT || 'food-data';
export const SEASON_WITH_FOOD_ENDPOINT =
  process.env.SEASON_WITH_FOOD_ENDPOINT || 'season-with-food';
export const SEASON_WITH_RECIPE_ENDPOINT =
  process.env.SEASON_WITH_RECIPE_ENDPOINT || 'season-with-recipes';
export const V1_ENDPOINT = process.env.V1_ENDPOINT || 'v1';
export const V2_ENDPOINT = process.env.V2_ENDPOINT || 'v2';
