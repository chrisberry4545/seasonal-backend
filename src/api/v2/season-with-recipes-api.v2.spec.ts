import {
  airtableMockSetup
} from '../mocks';
airtableMockSetup();

import {
  app
} from '../../app';

import supertest from 'supertest';
import { V2_ENDPOINT, SEASON_WITH_RECIPE_ENDPOINT } from '../../config';

const v2SeasonWithRecipesUrl = `${V2_ENDPOINT}/${SEASON_WITH_RECIPE_ENDPOINT}`;

describe('Get all seasons with recipes', () => {
  const makeAllSeasonWithRecipesRequest = () => {
    return supertest(app).get(`/${v2SeasonWithRecipesUrl}`);
  };

  test('Returns a status of 200', async () => {
    const result = await makeAllSeasonWithRecipesRequest();
    expect(result.status).toBe(200);
  });

  test('Returns a full list of season data', async () => {
    const result = await makeAllSeasonWithRecipesRequest();
    expect(result.body).toMatchSnapshot();
  });

  test('Returns the name of the seasons', async () => {
    const result = await makeAllSeasonWithRecipesRequest();
    expect(result.body[0].name).toBe('January');
    expect(result.body[1].name).toBe('February');
  });

  test('Populates the recipes in the seasons', async () => {
    const result = await makeAllSeasonWithRecipesRequest();
    expect(result.body[0].recipes).toHaveLength(1);
    expect(result.body[0].recipes[0].name).toBe('recipe 1');
  });
});

describe('Get single season with recipes', () => {
  const makeSingleSeasonWithRecipesRequest = (id: string = '0') => {
    return supertest(app).get(`/${v2SeasonWithRecipesUrl}/${id}`);
  };

  const makeSeasonRequestWithRecipes = () => {
    return makeSingleSeasonWithRecipesRequest('1');
  };

  test('Returns a status of 200', async () => {
    const result = await makeSingleSeasonWithRecipesRequest();
    expect(result.status).toBe(200);
  });

  test('Retrieves a single season', async () => {
    const result = await makeSingleSeasonWithRecipesRequest();
    expect(result.body).toMatchSnapshot();
  });

  test('Retrieves a single season with food and recipe data', async () => {
    const result = await makeSeasonRequestWithRecipes();
    expect(result.body).toMatchSnapshot();
  });

  test('Does not populate a seasons recipes if they do not exist', async () => {
    const result = await makeSingleSeasonWithRecipesRequest();
    expect(result.body.recipes).toBeUndefined();
  });

  test('Populates a seasons recipes if they exist', async () => {
    const result = await makeSeasonRequestWithRecipes();
    expect(result.body.recipes.length > 0).toBe(true);
  });

  test('Populates a seasons recipes isVegetarian if true', async () => {
    const result = await makeSeasonRequestWithRecipes();
    expect(result.body.recipes[0].isVegetarian).toBe(true);
  });

  test('Populates a seasons recipes isVegan if true', async () => {
    const result = await makeSeasonRequestWithRecipes();
    expect(result.body.recipes[1].isVegan).toBe(true);
  });

  test('Does not return any primaryFood on a recipe', async () => {
    const result = await makeSeasonRequestWithRecipes();
    expect(result.body.recipes[0].primaryFood).toBeUndefined();
  });

  test('Does not return any secondaryFood on a recipe', async () => {
    const result = await makeSeasonRequestWithRecipes();
    expect(result.body.recipes[0].secondaryFood).toBeUndefined();
  });

  test('Does not return any food', async () => {
    const result = await makeSeasonRequestWithRecipes();
    expect(result.body.food).toBeUndefined();
  });
});
