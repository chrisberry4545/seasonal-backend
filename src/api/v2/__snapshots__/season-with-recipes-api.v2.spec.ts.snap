// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get all seasons with recipes Returns a full list of season data 1`] = `
Array [
  Object {
    "id": "0",
    "name": "January",
    "recipes": Array [
      Object {
        "id": "r1",
        "isVegan": true,
        "name": "recipe 1",
      },
    ],
  },
  Object {
    "id": "1",
    "name": "February",
    "recipes": Array [
      Object {
        "id": "r1",
        "isVegan": true,
        "name": "recipe 1",
      },
    ],
  },
]
`;

exports[`Get single season with recipes Retrieves a single season 1`] = `
Object {
  "id": "0",
  "name": "January",
}
`;

exports[`Get single season with recipes Retrieves a single season with food and recipe data 1`] = `
Object {
  "id": "1",
  "name": "February",
  "recipes": Array [
    Object {
      "id": "r2",
      "isVegetarian": true,
      "name": "other recipe",
    },
    Object {
      "id": "r1",
      "isVegan": true,
      "name": "recipe 1",
    },
  ],
}
`;
